packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,251 @@
+   */
+  match(route: string): RouteTreeNodeMetadata | undefined {
+    let bestMatchMetadata: RouteTreeNodeMetadata | undefined;
+    let lowestIndex = Infinity;
+
+    for (const { insertionIndex, metadata } of this.traverseByRoute(route)) {
+      if (insertionIndex < lowestIndex) {
+        bestMatchMetadata = metadata;
+        lowestIndex = insertionIndex;
+      }
+    }
+
+    return bestMatchMetadata;
+  }
+
+  /**
+   * Converts the route tree into a serialized format representation.
+   * This method converts the route tree into an array of metadata objects that describe the structure of the tree.
+   * The array represents the routes in a nested manner where each entry includes the route and its associated metadata.
+   *
+   * @returns An array of `RouteTreeNodeMetadata` objects representing the route tree structure.
+   *          Each object includes the `route` and associated metadata of a route.
+   */
+  toObject(): RouteTreeObjectRepresentation {
+    return Array.from(this.traverse());
+  }
+
+  /**
+   * Constructs a `RouteTree` from an object representation.
+   * This method is used to recreate a `RouteTree` instance from an array of metadata objects.
+   * The array should be in the format produced by `toObject`, allowing for the reconstruction of the route tree
+   * with the same routes and metadata.
+   *
+   * @param value - An array of `RouteTreeNodeMetadata` objects that represent the serialized format of the route tree.
+   *                Each object should include a `route` and its associated metadata.
+   * @returns A new `RouteTree` instance constructed from the provided metadata objects.
+   */
+  static fromObject(value: RouteTreeObjectRepresentation): RouteTree {
+    const tree = new RouteTree();
+
+    for (const { route, ...metadata } of value) {
+      tree.insert(route, metadata);
+    }
+
+    return tree;
+  }
+
+  /**
+   * Creates an empty route tree node with the specified segment.
+   * This helper function is used during the tree construction.
+   * @param segment - The route segment that this node represents.
+   * @returns A new, empty route tree node.
+   */
+  private createEmptyRouteTreeNode(segment: string): RouteTreeNode {
+    return {
+      segment,
+      insertionIndex: -1,
+      children: new Map(),
+    };
+  }
+
+  /**
+   * A generator function that recursively traverses the route tree and yields nodes.
+   * This allows for easy iteration over the tree's nodes.
+   * @returns A generator that yields the metadata of each node in the tree.
+   */
+  private *traverse(): Generator<RouteTreeNodeMetadata> {
+    function* traverseNode(node: RouteTreeNode): Generator<RouteTreeNodeMetadata> {
+      if (node.metadata) {
+        yield node.metadata;
+      }
+
+      for (const [, childNode] of node.children) {
+        yield* traverseNode(childNode);
+      }
+    }
+
+    yield* traverseNode(this.root);
```

Comment:
Consider: You could use define this function with the node as a parameter which defaults to `this.root` and avoids the extra function definition altogether.

```typescript
private *traverse(node: RouteTreeNode = this.route): Generator<RouteTreeNodeMetadata> {
  if (node.metadata) yield node.metadata;

  for (const [, childNode] of node.children) {
    yield* this.traverseNode(childNode);
  }
}
```